src/transformers/models/modernbert/modular_modernbert.py

```diff
@@ -0,0 +1,1515 @@
+        self.hidden_activation = hidden_activation
+        self.global_attn_every_n_layers = global_attn_every_n_layers
+        self.local_attention = local_attention
+        self.local_rope_theta = local_rope_theta
+        self.skip_first_prenorm = skip_first_prenorm
+        self.embedding_norm = embedding_norm
+        self.embedding_dropout = embedding_dropout
+        self.mlp_bias = mlp_bias
+        self.mlp_dropout = mlp_dropout
+        self.unpad_inputs = unpad_inputs
+        self.unpad_no_grad = unpad_no_grad
+        self.decoder_bias = decoder_bias
+        self.classifier_dropout = classifier_dropout
+        self.classifier_pooling = classifier_pooling
+        self.classifier_bias = classifier_bias
+        self.classifier_norm = classifier_norm
+        self.classifier_activation = classifier_activation
+        self.deterministic_flash_attn = deterministic_flash_attn
+        self.sparse_prediction = sparse_prediction
+        self.sparse_pred_ignore_index = sparse_pred_ignore_index
+
+        if unpad_inputs is None:
+            self.unpad_inputs = self._attn_implementation == "flash_attention_2"
+
+
+class ModernBertModuleType(str, Enum):
+    in_module = "in"
+    out_module = "out"
+    embedding = "emb"
+    final_out = "final_out"
+
+
+class ModernBertPoolingType(str, Enum):
+    cls = "cls"
+    mean = "mean"
+    max = "max"
+
+
+# Copyright 2023 OLMo Authors
+# License: Apache-2.0
+
+
+def _init_modernbert_weights(
+    config: ModernBertConfig,
+    module: Union[nn.Linear, nn.Embedding],
+    module_type: ModernBertModuleType,
+) -> None:
+    """
+    Initialize weights of a linear or embedding module.
+
+    :param config: The model config.
+    :param module: The linear or embedding submodule to initialize.
+    """
+    if module_type is None:
+        raise RuntimeError("When using the full megatron init, every module must have a type.")
+
+    cutoff_factor = config.initializer_cutoff_factor
+    if cutoff_factor is None:
+        cutoff_factor = 3
+
+    if module_type == ModernBertModuleType.in_module:
+        std = config.initializer_range  # for att_proj (same as QKV), ff_proj
+    elif module_type == ModernBertModuleType.out_module:
+        std = config.initializer_range / math.sqrt(2.0 * config.num_hidden_layers)  # for attn_out, ff_out
+    elif module_type == ModernBertModuleType.embedding:
+        std = config.initializer_range  # token embeddings (wte)
+    elif module_type == ModernBertModuleType.final_out:
+        std = config.hidden_size**-0.5  # final output (ff_out)
+    else:
+        raise RuntimeError(f"Unknown module type '{module_type}'")
+
+    nn.init.trunc_normal_(
+        module.weight,
+        mean=0.0,
+        std=std,
+        a=-cutoff_factor * std,
+        b=cutoff_factor * std,
+    )
+
+    if isinstance(module, nn.Linear):
+        if module.bias is not None:
+            nn.init.zeros_(module.bias)
+
+
+def _unpad_modernbert_input(
```

Comment:
correct me if I am wrong, the inputs are the same as any other LLMs no? In that case if you want to unpad you should be using https://github.com/huggingface/transformers/blob/1b6cb1eefc140ea2dbd74e23389050ebcb7f417d/src/transformers/modeling_flash_attention_utils.py#L63-L69 available in the utils!

Reply:
I'm afraid not quite. We're unpadding much earlier and repadding much later, so that e.g. even the MaskedLM can take advantage of it. As a result, `_upad_input` (and `_flash_attention_forward`) aren't viable here.